### DISCUSSION TITLE
pydeck: Load a local .glb file in ScengraphLayer

### DISCUSSION THREAD

**Discussion by harisbal** [visgl/deck.gl#6323] [Q&A] [answered]
I was wondering whether it is possible to load a local .glb file in my Scenegraph Layer using pydeck.
The following code does not seem to work (I'm using jupyter)

```python
import pydeck as pdk
layer = pdk.Layer(
    "ScenegraphLayer",
    data=[{'coordinates': [-122.466233, 37.684638]}],
    scenegraph=r'./BoxAnimated.glb',
    get_position="coordinates",
)

view_state = pdk.ViewState(latitude=37.684638, longitude=-122.466233, zoom=18, bearing=0, pitch=0)

r = pdk.Deck(layer, initial_view_state=view_state, map_style="light")
r.show()
```
Am I missing something?
Cheers

**Comment by kylebarron** [comment]
You should be able to look at the JavaScript console for errors. I haven't tried to use pydeck with the ScenegraphLayer, but I'm guessing this doesn't work because Javascript isn't able to load files in your local directory from the browser. The example loads data from a remote URL: https://github.com/visgl/deck.gl/blob/cb9f30700387b9726c574fb899f5acbb7b9b701a/bindings/pydeck/examples/scenegraph_layer.py#L16

One easy way to get around this is to start a local server in the directory with your files. You can run `python -m http.server` to start a local server (by default on port 8000), and then load the scenegraph data from `http://localhost:8000/BoxAnimated.glb`.

**Comment by harisbal** [comment]
Thank you @kylebarron 
Pydeck seems to ignore the `scenegraph` parameter since I do not see any errors in the JavaScript console even when I try to load files that do not exist.

Running a local server could be an option but for the moment it raises a CORS error. In addition my end goal is to include the `scenegraph` inside a flask app therefore I am not sure this solution could help.

I know that I could upload the file and read it from a remote URL but the size of the actual file I need to open (>100mb) makes that quite difficult

**Comment by kylebarron** [comment]
> Pydeck seems to ignore the `scenegraph` parameter since I do not see any errors in the JavaScript console even when I try to load files that do not exist

I'm not familiar with that layer, but you should consult the new example @ajduberstein added last week in https://github.com/visgl/deck.gl/pull/6308

**Comment by ajduberstein** [comment, marked as answer]
@harisbal just published a version of pydeck with the ScenegraphLayer changes, so you might want to upgrade (`pip install -U pydeck` if you're using pip). The Python SimpleHTTPServer doesn't enable CORS by default. Try `http-server` for this.

```bash
npm install -g http-server
# In the directory of your .glb file
http-server -p 8000 --cors
```

I can verify that this code works for me (from the `examples/` directory in pydeck):

```python
import pandas as pd

from pydeck import Layer, Deck
from pydeck.data_utils import compute_view
from pydeck.types import String

DATA_URL = "https://raw.githubusercontent.com/visgl/deck.gl-data/master/website/bart-stations.json"
# Referencing the local file server
SCENEGRAPH_URL = "http://localhost:8000/BoxAnimated.glb"

# Automatically fit viewport location based on data
df = pd.read_json(DATA_URL)
view = compute_view(df["coordinates"])

layer = Layer(
    type="ScenegraphLayer",
    id="scenegraph-layer",
    data=DATA_URL,
    pickable=True,
    scenegraph=SCENEGRAPH_URL,
    get_position="coordinates",
    get_orientation=[0, 180, 90],
    size_scale=500,
    _animations={"*": {"speed": 5}},
    _lighting=String("pbr"),
)

# Render
r = Deck(layer, initial_view_state=view)
r.to_html("scenegraph_layer.html")
```

Give this a try and let us know if it works.

**Reply by harisbal** [reply]
It works great! Thank you very much @adriandarian and @kylebarron

**Reply by harisbal** [reply]
Following up our discussion @ajduberstein I would like to ask whether loading a local `.glb` file using `pydeck` is something to be implemented in the future or whether running an external server would always be required.
Cheers

**Reply by ajduberstein** [reply]
@harisbal open to this if you want to file an issue

**Reply by bobfortin64** [reply]
I am able to load local files by disabling CORS when starting chrome: "C:\Program Files (x86)\Google\Chrome\Application\chrome.exe"  --user-data-dir="C://chrome-dev-disabled-security" --disable-web-security --disable-site-isolation-trials

**Reply by muntakim1** [reply]
How to reduce the hight? The objects show in 10x above of the map?

**Reply by iimmer** [reply]
@ajduberstein can this examle run with pydeck? https://deck.gl/examples/scenegraph-layer

**Comment by Pessimistress** [comment]
Moving this thread to Discussion. Please only use templates to open issues.
